### ISSUE TITLE
Regex in rule 912150 (DOS counter) correct?

### ISSUE THREAD

**Issue by jschleus** [SpiderLabs/owasp-modsecurity-crs#470]
Although I just detected the recently opened issue "DOS counter broken for files without dots #462" I  opened a new issue since it has a somewhat different aspect (but it maybe something related):

1) The small problem (or bug?) I want to report regards the regex used within the rule 912150 in the file REQUEST-912-DOS-PROTECTION.conf

 `SecRule REQUEST_BASENAME "\.(.*)$" \`

that matches a maximum and not a minimum string so for e.g. "foo.bar.png" would match "bar.png" but not as probably expected the pure extension "png". So an expression like

 `"\.([^\.]*)$"`

or more specific (regarding only the current entries in tx.static_resources in modsecurity_crs_10_setup.conf.example) like

 `"\.([a-zA-Z]{2,4})$"`

seems more appropriate.

Or one can put the the extension separating "." into the the capturing parentheses and use a more general rule (allowing for e.g. also extensions like ".c" and ".mp3") like

 `"(\.[a-zA-Z0-9]{1,4})$"`

so also the helping construct with the limiting "/" can be ommitted and the variable tx.static_resources can be simplified to

  `".jpg .jpeg .png .gif .js .css .ico .svg .webp"`

if at the same time the "/" is replaced by a space in

  `setvar:tx.extension=%{TX.1} ,\`
  `chain"`
  `SecRule TX:EXTENSION "!@within %{tx.static_resources} " \`

2) And here only some small additional less important remarks (from a CRS layman):

The last line of the rule 912150

       setvar:ip.dos_counter=+1"

maybe lack a double quote (") in front of setvar (but it seems to work with and without it)?

Also it seems not simple to detect non-static resources by just excluding some static resources identified by the requested filename (extensions). Perhaps an inverse method could be simpler trying to count non-static resources by explicitly including them (PHP SQL ... ?). In addition DOS attacks can surely also be done using some static resources so the variable name tx.static_resources (currently only used in REQUEST-912-DOS-PROTECTION.conf) seems not optimal to me (something like tx.dos_uncounted_resources maybe slightly better).

Ok, although REQUEST-912-DOS-PROTECTION.conf is now much better documented (thanks!) I have yet not completely understood how it works (I must learn more about modsecurity and CRS). But in the sentence under "Description of mechanics:"

  `# If the counter (IP:DOS_COUNTER) hits a limit (TX:DOS_COUNTER_THRESHOLD), then a burst is identified ...`

"hits" seems to me not the correct word and "exceeds" the better word (since the comparison operator is @gt and not @ge).

Since the burst counter is simply reset to 0 I assume still active attackers get shortly access again (also if they send further attacking requests within the block period).

Also a small unimportant typo I found (resoures -> resources):

  `# DOS Counter: Count the number of requests to non-static resoures`

By the way I currently do first steps to experiment with two variations of REQUEST-912-DOS-PROTECTION.conf in order to protect my server against "mirroring" (downloading all or many of the offered often big files) by counting downloaded files (per IP; yes, that files are principally identifiable by file extension) and other "misuses" (shortly repeated downloads and also many partial downloads of a single file with a download volume exceeding the original file size by a factor of 3-30; maybe caused by erroneous browsers or ruthless download managers?) by counting bytes (per IP). I do that til now with some success asynchronously analyzing periodically the Apache log files with a script but hope that a modsecurity solution will work similar but react faster (although it seems less flexible and may cost more resources).

**Comment by csanders-git** [comment]
These are some good points I didn't think of while testing. I'll be investigating them ASAP

**Comment by csanders-git** [comment]
Your regex here ```(\.[a-zA-Z0-9]{1,4})$``` actually seems perfect. I'm testing it now to make sure it performs correctly but this takes into account the issue you discussed. Additionally, I've made some of the other changes you described. 

I agree it is possible to use static resources to do a DOS, in many cases people run caching servers before ModSecurity and the rule as implemented does make sense from a performance perspective. However, I think that your idea is very valid.

I think we should open a feature requests for these types of ideas. While we're coming to a close on RC1 functionality certainly such functionality can make it into 3.1 as it increases the underlying protection. 

May I also say you've done a wonderful job spotting and testing this. Thank you.

**Comment by csanders-git** [comment]
you are right about the double quote, Apache tries to help with this after line continuations is my understanding but this is bad form. I added the quote back in. I think the rule seems backwards but otherwise you are stuck with a skip, where u have to skip the secaction to add to the score. I am not sure which is more performant, possibly some research for while this is in RC1 :). I ended up going similar to your regex but keeping the original intent ```.*?(\.\w{1,4})?$``` . I also kept the slash separator even though you may be right going {3,4} may remove the need for that.

You knowledge of ModSecurity is very good. Please continue to test RC1, we do appreciate it and ANY feedback is helpful. Much of it will make its way into RC2 or 3.1 :)

**Comment by csanders-git** [comment]
So the issues I think we should split out for 3.1 consideration
  - static file based counter
  - Automate "mirroring" protection.

@jschleus does that sound right?

**Comment by jschleus** [comment]
Yes, that is completely ok for me.

By the way, since some  "legal" clients use partial GET requests with relative small byte ranges (for whatever reason)  they must issue also plenty of such requests to fetch a large file in total so the chance is great that they are blocked by  REQUEST-912-DOS-PROTECTION.conf (what maybe a kind of an FalsePositives). On the other side (as I mentioned already above) I see some other clients "misusing"  partial GET requests downloading a multiple of the original file size. 

Motivated by your nice words I just tried to find a first approach for a solution that avoids the FPs but  blocks the nasty ones. Waiting for a  RESPONSE_STATUS of 206 and counting the delivered bytes for e.g. via RESPONSE_CONTENT_LENGTH seems too late so the  first idea was to check REQUEST_HEADERS:Range and counting the requested bytes but that was already hard for me since I didn't know how to build the difference between two modsecurity variables (I solved it using a helping variable) but the real problem begins if the client specifies more complex expressions (especially  with an absent last-byte position to request the remainder of a file). So I got the impression only an external script could help and coded my first lines in Lua but without suited knowledge that was an frustrating job. But after many trial and errors I got now at least principally the file size of the requested file respectively the expected response content length assigned to an countable TX.variable. A first step but ...
